api/python/provisioner/ui/form_window.py

```diff
@@ -0,0 +1,202 @@
+#
+#
+import curses
+import config
+from window import Window
+from color_code import ColorCode
+from text_box import TextBox
+from success import SuccessWindow
+from pillar import Pillar
+
+
+class FormWindow(Window):
+
+    data = {
+               'Ip': {
+                         'default': '127.0.0.1',
+                         'validation': 'ipv4',
+                         'pillar_key': 'srvnode-0/ip'
+                     }
+           }
+    component_type = 'Default User input'
+
+    def update_pillar(self):
+        result = True
+        for key in self.data.keys():
+            if self.data[key]['pillar_key']:
+                if (not Pillar.set_pillar(self.data[key]['pillar_key'],
+                                          self.data[key]['default'])):
+                    result = False
+        return result
+
+    def format_data(self):
+        result = ''
+        for key in self.data.keys():
+            result = result + f"{key}: {self.data[key]['default']} "
+        return result
+
+    def submit_button(self, x, y, menu, selected_rows):
+        buttons = {
+           "Submit": x + 15,
+           "Cancel": x + 5
+        }
+        selected_button = None
+
+        if selected_rows == len(menu):
+            selected_button = "Submit"
+        elif selected_rows == len(menu) + 1:
+            selected_button = "Cancel"
+
+        for button in buttons.keys():
+            if selected_button == button:
+                cod = ColorCode.get_color_pair(1)
+                self.on_attr(cod)
+                self._window.addstr(y + 1, buttons[button], button)
+                self.off_attr(cod)
+            else:
+                self._window.addstr(y + 1, buttons[button], button)
+
+    @staticmethod
+    def parse_input(kwargs):
+        if 'selected' not in kwargs:
+            selected_rows = 0
+        else:
+            selected_rows = kwargs['selected']
+
+        if 'edit' not in kwargs:
+            edit = False
+        else:
+            edit = kwargs['edit']
+        return selected_rows, edit
+
+    def action(self):
+        pass
+
+    def create_window(self, **kwargs):
+        color_code = kwargs['color_code']
+        self._parent = kwargs['component']
+
+        selected_rows, edit = FormWindow.parse_input(kwargs)
+        self._window.border()
+        col_code_attr = ColorCode.get_color_pair(color_code)
+        self.create_menu_head()
+
+        x = self.get_max_width() // 2
+        y = self.get_max_height() // 3 - 1
+        self.enable_keypad()
+
+        self.on_attr(col_code_attr)
+        self._window.addstr(y, 3, f"Please enter {self.component_type}"
+                            " for this machine")
+        self.off_attr(col_code_attr)
+
+        values = list(self.data.keys())
+        count_m = len(values)
+        x = 6
+
+        for idx, row in enumerate(values):
+            y = self.get_max_height() // 3 - count_m // 2 + (idx + 1) * 2
```

Comment:
why we evaluate the same multiple times (`self.get_max_height() // 3`)? Is there a chance that the result may change in the middle. If now - please use some well-named variable instead.

Reply:
added variable in place of this

Reply:
addressed